### ISSUE TITLE
Organize model classes

### ISSUE THREAD

**Issue by hayribakici** [rinukkusu/spotify-dart#172]
I figured out, that the model classes are very messy with many dupicates, which makes this library quite error prone.

In order to give more clarity, I am suggesting to model the library as shown in this diagram:

[![](https://mermaid.ink/img/pako:eNqtWNtu2zwMfhXDwO6aPYARDGh3wAa0Q7H0vxlyo1hqIkSWDEluZ-Svn32UTxEV2fGw-cqiKZL6eJRPaa4oS7N0tVptpeVWsCwxpbL8uV5Rom1SwHeR5IIYk1BO9poUW9lyb2VL_TQQE3g23daPSlom7R0xLFn_v1oF9Hnej8SyvdJ1jKvjuK0oV3dKHTe8KAXrGANiwDkt7FGQWnBjfVmYhvmmJW0O6nXGaLGrCmTwmeBxdN_evUseVVkJormtk_XFxx_MWM1zy5X8wYhRMlm_f-_zxCx40iRHkE0KuuD0CJ6smcNqG0DqU3yeaRmfS24g-K6Yirj6RR9HB1Japn0jEAlxuXB2HF2gR0Lz1H13z3p9uwNTSG4_fDhTVwmnWbIBE-Xepx40e8b0tzllSxRVmsc02bqE7O3lPMEiqi7Mk5MvglT2oLTJkntwTXPbrhrE8EK4IDvBHog-MjtwdivEmauy1nx_sM6k_tX_TpnJNS-dOyfIX20h_E-M8pCb_QLfyf-0MMgRBdmzwbRvboEsE0TuK8fhEwuQTs6gdURJimCtNbFK96K_90sk3SpLRB9VYAMfyt2UG7AD8nFjqwEFbBMX5JWR05yr_hq3EA3NBOQh-wT1GnkQ1o-a5dyMzoqZHJw7FiKdlUR8o8jMPZMaO68cayW2bywYcTP8knSKKQ4BOiMQFYWFPCsh1Cv4ct72K6CHRwsUD80Sq-a5kkv8GApD5fH0J3vCMgLBvYPgzsIOHOweCvblboigF85ewQN_VzRoBSkK9Icg_kvBc26XOP2Kg4Ztov6qjGUUOQ9mBpeBV6vPRCr15EhGjRFeFexRjUVmIvA7soH5JPOmlMAXwdATpKeAUqIclC9sgUP-SX1Rr9I1Zg_OagdeQ0eSpIRjodO7hnk091weZ4-JD9ht6k1rh5t4tXXoLS-z-T9qfJPRGiI8F4zxDheC3mJsDhj4tqf1uYr19bQeuJ4lDh0aIjGEbfUcp4521SzFmHKTf9dLUj6e3hNNJpa6E3nFzb3KCfKYgOhj9ItWRRSKFokAA9cTM3-ED-P2fB9AI-ID_8Ulng_PPWN6_vjCiK00o_242Dw1SybPghkDaRubPvNOUJY8BfoCTfg2NXxsBkIzs2lodiMvLllNkKrk5U-PRyhl9BbOEBbiycPhmhLzqxN5V8cQ6-MmS6BTiqhc78pjrnpdt5ei7PKedE10f5tC8j9LaCxdGvlaijb_zutSK1rlHuGyVFXyKKGSB0b4zvFvp46M6u8FGcePfw-9EOmKdbfVvQXh4d2TFh2dnDPTPfasuAUC_R1wjxn7bIuLf01tUUhvUlBTEE7TLG0N2Kb2wKAapxm8UkB6m27lG_DBrUxtapmn2TMRht2kVelGgv7Py0h19yOlH7ofOu1_nZu0JPKnUgPP228uDohq?type=png)](https://mermaid.live/edit#pako:eNqtWNtu2zwMfhXDwO6aPYARDGh3wAa0Q7H0vxlyo1hqIkSWDEluZ-Svn32UTxEV2fGw-cqiKZL6eJRPaa4oS7N0tVptpeVWsCwxpbL8uV5Rom1SwHeR5IIYk1BO9poUW9lyb2VL_TQQE3g23daPSlom7R0xLFn_v1oF9Hnej8SyvdJ1jKvjuK0oV3dKHTe8KAXrGANiwDkt7FGQWnBjfVmYhvmmJW0O6nXGaLGrCmTwmeBxdN_evUseVVkJormtk_XFxx_MWM1zy5X8wYhRMlm_f-_zxCx40iRHkE0KuuD0CJ6smcNqG0DqU3yeaRmfS24g-K6Yirj6RR9HB1Japn0jEAlxuXB2HF2gR0Lz1H13z3p9uwNTSG4_fDhTVwmnWbIBE-Xepx40e8b0tzllSxRVmsc02bqE7O3lPMEiqi7Mk5MvglT2oLTJkntwTXPbrhrE8EK4IDvBHog-MjtwdivEmauy1nx_sM6k_tX_TpnJNS-dOyfIX20h_E-M8pCb_QLfyf-0MMgRBdmzwbRvboEsE0TuK8fhEwuQTs6gdURJimCtNbFK96K_90sk3SpLRB9VYAMfyt2UG7AD8nFjqwEFbBMX5JWR05yr_hq3EA3NBOQh-wT1GnkQ1o-a5dyMzoqZHJw7FiKdlUR8o8jMPZMaO68cayW2bywYcTP8knSKKQ4BOiMQFYWFPCsh1Cv4ct72K6CHRwsUD80Sq-a5kkv8GApD5fH0J3vCMgLBvYPgzsIOHOweCvblboigF85ewQN_VzRoBSkK9Icg_kvBc26XOP2Kg4Ztov6qjGUUOQ9mBpeBV6vPRCr15EhGjRFeFexRjUVmIvA7soH5JPOmlMAXwdATpKeAUqIclC9sgUP-SX1Rr9I1Zg_OagdeQ0eSpIRjodO7hnk091weZ4-JD9ht6k1rh5t4tXXoLS-z-T9qfJPRGiI8F4zxDheC3mJsDhj4tqf1uYr19bQeuJ4lDh0aIjGEbfUcp4521SzFmHKTf9dLUj6e3hNNJpa6E3nFzb3KCfKYgOhj9ItWRRSKFokAA9cTM3-ED-P2fB9AI-ID_8Ulng_PPWN6_vjCiK00o_242Dw1SybPghkDaRubPvNOUJY8BfoCTfg2NXxsBkIzs2lodiMvLllNkKrk5U-PRyhl9BbOEBbiycPhmhLzqxN5V8cQ6-MmS6BTiqhc78pjrnpdt5ei7PKedE10f5tC8j9LaCxdGvlaijb_zutSK1rlHuGyVFXyKKGSB0b4zvFvp46M6u8FGcePfw-9EOmKdbfVvQXh4d2TFh2dnDPTPfasuAUC_R1wjxn7bIuLf01tUUhvUlBTEE7TLG0N2Kb2wKAapxm8UkB6m27lG_DBrUxtapmn2TMRht2kVelGgv7Py0h19yOlH7ofOu1_nZu0JPKnUgPP228uDohq)

The main thing here are the two classes `SpotifyContentBase` and `SpotifyContent`, which all of the content classes (`Tracks`, `Playlist`, `Album`, `Episodes`, `Show`, `Artist`, the last one is really debatable). The idea behind this is that every class that has the attributes `id`, `type`, `href` and `uri` is considered a `SpotifyContent`.

Furthermore, I am suggesting to add some "model building blocks" in the form of `mixin`s where model classes can just implement them and include some model properties. Furthermore, this could be also useful, if  the `@JsonKey` needs a `fromJson` function, so that code is not copied around. This could look like as follows:

```dart
class Foo with BuildingBlock {
  
  @JsonKey(name: 'some_property')
  int someProperty;
}

mixin BuildingBlock {

  static String _someMethod(dynamic json) => 'Hi';
  
  String? _inheritedProperty;

  @JsonKey(fromJson: _someMethod)
  String? get inheritedProperty => _inheritedProperty;
}
```
In order to properly inherit the attribute, it is important to wrap it in a property. Concretely, mixins such as `Copyright`, `ExternalUrls`, `AvailableMarkets` and `Popularity` (as discussed in #138) could be one of those (to name a few) building blocks.

What do you think?

**Comment by rinukkusu** [comment]
Thanks for the writeup and input - will take an indepth look at this hopefully this weekend!

**Comment by hayribakici** [comment]
Thank you!

**Comment by rinukkusu** [comment]
I completely agree - it's a little messy to work with and reason about. The current implementation is a naive 1:1 mapping of the Spotify API models from the documentation with few abstractions. This is also from a time, where mixins etc were not a thing. 😆 

Your mockup of the the class design seems pretty reasonable - you took the time to go through every model and checked what field can be extracted, very nice!

Regarding the building blocks section: this is what you mean with the `mixin` "type" in your diagram, right?

**Comment by hayribakici** [comment]
Yes, exactly. Sorry, the naming wasn't very consistent. Thank you, however, for taking the time to go through the mockup

**Comment by hayribakici** [comment]
I tried the above approach with the `mixin`s on my local branch. Unfortunately, the concept does not work. `jsonserializer` does not generate the necessary code for the attributes from the mixin. Meaning, a class

```dart
class Album with Popularity {
  String? name;
}

mixin Popularity {
  int? popularity;
}
```
wont generate the code for the attribute `popularity` and fill it with a value accordingly. Just wanted to share my insights.
